Core/include/Acts/TrackFinding/MeasurementSelector.hpp

```diff
@@ -106,60 +107,81 @@ class MeasurementSelector {
-    for (auto& trackState : candidates) {
-      // Take the parameter covariance
-      // const auto predicted = tackState.predicted();
-      // const auto predictedCovariance = trackState.predictedCovariance();
-
-      double chi2 = calculateChi2(
-          // This abuses an incorrectly sized vector / matrix to access the
-          // data pointer! This works (don't use the matrix as is!), but be
-          // careful!
-          trackState
-              .template calibrated<MultiTrajectoryTraits::MeasurementSizeMax>()
-              .data(),
-          trackState
-              .template calibratedCovariance<
-                  MultiTrajectoryTraits::MeasurementSizeMax>()
-              .data(),
-          trackState.predicted(), trackState.predictedCovariance(),
-          trackState.projector(), trackState.calibratedSize());
-
-      trackState.chi2() = chi2;
-
-      // Search for the measurement with the min chi2
-      if (chi2 < minChi2) {
-        minChi2 = chi2;
-        minIndex = index;
-      }
-
-      index++;
+    auto trackStateIterEnd = candidates.end();
+    {
+       auto trackStateIter = candidates.begin();
+       // Loop over all measurements to select the compatible measurements
+       // Sort track states which do not satisfy the chi2 cut to the end.
+       // When done trackStateIterEnd will point to the first element that
+       // does not satisfy the chi2 cut.
+       assert(trackStateIter != trackStateIterEnd);
+       for (;;) {
+          // Take the parameter covariance
+          // const auto predicted = tackState.predicted();
+          // const auto predictedCovariance = trackState.predictedCovariance();
+
+          double chi2 = calculateChi2(
+            // This abuses an incorrectly sized vector / matrix to access the
+            // data pointer! This works (don't use the matrix as is!), but be
+            // careful!
+            trackStateIter
+                ->template calibrated<MultiTrajectoryTraits::MeasurementSizeMax>()
+                .data(),
+            trackStateIter
+                ->template calibratedCovariance<
+                    MultiTrajectoryTraits::MeasurementSizeMax>()
+                .data(),
+            trackStateIter->predicted(), trackStateIter->predictedCovariance(),
+            trackStateIter->projector(), trackStateIter->calibratedSize());
+
+          trackStateIter->chi2() = chi2;
+
+          // only consider track states which pass the chi2 cut
+          if (chi2 >= maxChi2Cut || chi2 >= VariableCut<traj_t>(*trackStateIter, cuts, chi2CutOff, logger) ) {
+             --trackStateIterEnd;
+             // still check whether the element has the smallest chi2 in case an outlier is returned.
+             if (chi2<minChi2) {
+                minChi2 = chi2;
+                // the current element will be swapped with the last unchecked element if they are
+                // different
+                minIndex = trackStateIterEnd-candidates.begin();
```

Comment:
I didn't know that `vector` iterators support this. Maybe still better to use `std::distance`?